botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1027 @@
+    }
+    """
+
+    def __init__(self, rules, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.rules = [
+            RuleCreator.create(
+                assignments=self.assignments, provider=self.provider, **rule
+            )
+            for rule in rules
+        ]
+
+    def evaluate(self):
+        """If a tree rule's conditions evaluate successfully, iterate over its
+        subordinate rules and return a result if there is one. If any of the
+        subsequent rules are trees, the function will recurse until it reaches
+        an error or an endpoint rule.
+
+        :returns: An endpoint or error object.
+        """
+        if self.evaluate_conditions() is True:
+            for rule in self.rules:
+                evaluation = rule.evaluate()
+                if evaluation:
+                    return evaluation
+        return None
+
+
+class RuleCreator(Enum):
+    """A helper class that creates rule objects."""
+
+    endpoint = EndpointRule
+    error = ErrorRule
+    tree = TreeRule
+
+    @classmethod
+    def create(cls, **kwargs):
+        """Given a rule dictionary, locate the `type` attribute, using it to
+        determine the type of rule object to create. Then create and return it.
+
+        :type kwargs: dict
+        :params kwargs: An arbitrary dictionary containing required parameters
+        for different types of rule objects i.e. conditions, rules, errors,
+        endpoints etc.
+
+        :returns: A TreeRule/EndpointRule/ErrorRule object
+        """
+        rule = getattr(cls, kwargs.pop('type'))
+        return rule.value(**kwargs)
+
+
+class ParameterDefinition:
+    """A class that houses the spec of an individual parameter defined in a
+    rule set object i.e.
+    {
+        "Region": {
+            "builtIn": "AWS::Region",
+            "required": true,
+            "type": "String"
+        }
+    }
+    """
+
+    def __init__(
+        self,
+        name,
+        type,
+        documentation=None,
+        builtIn=None,
+        default=None,
+        required=None,
+        deprecated=None,
+    ):
+
+        self.name = name
+        self.type = type
+        self.documentation = documentation
+        self.built_in = builtIn
+        self.default = default
+        self.required = required
+        self.deprecated = deprecated
+
+    class ParameterType(Enum):
+        """An enum that translates a parameter defintion's `type` attribute to
+        its corresponding python native type.
+        """
+
+        String = str
+        Boolean = bool
+
+    def validate_input(self, input_param):
+        """Given an input parameter, validate that it matches the rules provided in
+        the spec.
+
+        :type input_param: Any
+        :param input_param: A value for a parameter provided by a client.
+
+        :raises: Exception if the input is of the incorrect type
+        """
+
+        correct_param_type = getattr(self.ParameterType, self.type).value
```

Comment:
The `type` parameter is provided at initialization, so it should be constant for the lifecycle of this class. Can we not resolve it to the correct Python type at that point instead of doing it every time `validate_input` is called?

Reply:
oh i see. yeah i'll move it